### PR TITLE
[Searchcache, webapp] Create Caching for WPT Metadata

### PR DESCRIPTION
**Pull request by KyleJu**
Fix for #1776 in searchcache and webapp. This change polls Metadata update in the background in searchcache and update Metadata through memcache in webapp on a 10-minute interval.

### CODE REVIEW

**Comment by KyleJu** [comment]
@Hexcles One question I have is whether I put the polling in [webapp](https://github.com/web-platform-tests/wpt.fyi/blob/master/webapp/web/main.go) or searchcache

**Comment by Hexcles** [comment]
Sorry for the late reply.

Both webapp and searchcache need metadata, but only searchcache has a long running daemon where you can do the polling, so we'd need to have a polling loop in a goroutine in searchcache that periodically updates the metadata tarball in memcache, and the webapp should then read from memcache.

File: shared/metadata_util.go [hunk 1]
```diff
@@ -14,17 +14,34 @@ import (
 	"strings"
 )
 
-const metadataArchiveURL = "https://api.github.com/repos/web-platform-tests/wpt-metadata/tarball"
+// metadataMapCached is the local cache of Metadata.
+var metadataMapCached map[string][]byte = nil
```
Comment (Hexcles) [inline shared/metadata_util.go]: We not only need this in-memory cache, but more importantly need to put it into memcache for use by webapp.

**Comment by Hexcles** [comment]
Actually, I just realized that memcache isn't available in AppEngine Flex: https://cloud.google.com/appengine/docs/flexible/go/migrating#memcache

So we need two different caching strategies:
* In webapp, we need to cache the metadata tarball in memcache and fetch a new one upon expriy.
* In searchcache, what you have now (polling + in-memory cache) should be good.

File: api/query/cache/service/main.go [hunk 2]
```diff
@@ -313,6 +313,12 @@ func main() {
 	// after backfilling was started.
 	go poll.KeepRunsUpdated(store, logger, *updateInterval, *updateMaxRuns, idx)
 
+	var netClient = &http.Client{
+		Timeout: time.Second * 5,
+	}
+	// Polls Metadata update every 10 minutes.
+	go shared.KeepMetadataUpdated(netClient, logger, time.Second*10)
```
Comment (Hexcles) [inline api/query/cache/service/main.go]:
```suggestion
	go shared.KeepMetadataUpdated(netClient, logger, time.Minute*10)
```

Also this probably shouldn't live in `shared` as it's only used by searchcache.

File: shared/poll.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	"net/http"
+	"time"
+)
+
+// KeepMetadataUpdated implements updates to metadataMapCached via simple polling every
+// interval duration.
+func KeepMetadataUpdated(client *http.Client, logger Logger, interval time.Duration) {
+	// Start by waiting polling interval.
+	logger.Infof("Starting Metadata update via polling; waiting polling interval first...")
+	time.Sleep(interval)
```
Comment (Hexcles) [inline shared/poll.go]: Why does it start by waiting first? Shall we fetch metadata immediately instead?

**Comment by KyleJu** [comment]
> Actually, I just realized that memcache isn't available in AppEngine Flex: https://cloud.google.com/appengine/docs/flexible/go/migrating#memcache
> 
> So we need two different caching strategies:
> 
> * In webapp, we need to cache the metadata tarball in memcache and fetch a new one upon expriy.
> * In searchcache, what you have now (polling + in-memory cache) should be good.

Per our discussion offline, metadata caching for standard webapp will be done using memcache in a separate PR. This is ready for review

File: shared/poll.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	logger.Infof("Metadata update via polling started")
+
+	for {
+		start := time.Now()
+		err := UpdatedMetadata(client, logger)
+		if err != nil {
+			logger.Errorf("Error fetching Metadata for update: %v", err)
+		}
+
+		wait(start, interval)
```
Comment (Hexcles) [inline shared/poll.go]: I think this is a bit overly complicated. We can simply `time.Sleep(interval)` again and disregard how much time the download takes.

File: api/metadata_cache.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+func (f webappMetadataFetcher) Fetch() (res map[string][]byte, err error) {
+	metadataMap, err := getMetadataFromMemcache(f.ctx, f.log, f.client, f.url)
+	if err == nil && metadataMap != nil {
+		return metadataMap, nil
+	}
+
+	return shared.CollectMetadataWithURL(f.client, f.url)
+}
+
+func getMetadataFromMemcache(ctx context.Context, log shared.Logger, client *http.Client, url string) (res map[string][]byte, err error) {
```
Comment (KyleJu) [inline api/metadata_cache.go]: Note to the reviewer: This is taken from #1784 and have tested separately

**Comment by KyleJu** [comment]
@Hexcles ready for review

File: api/metadata_handler.go [hunk 6]
```diff
@@ -35,8 +34,8 @@ func apiMetadataHandler(w http.ResponseWriter, r *http.Request) {
 	ctx := shared.NewAppEngineContext(r)
 	client := shared.NewAppEngineAPI(ctx).GetHTTPClient()
 	logger := shared.GetLogger(ctx)
-	metadataURL := shared.MetadataArchiveURL
-	delegate := MetadataHandler{logger, client, metadataURL}
+	fetcher := webappMetadataFetcher{ctx: ctx, client: client, log: logger, url: shared.MetadataArchiveURL}
+	delegate := MetadataHandler{logger, fetcher}
 
 	// Serve cached with 5 minute expiry. Delegate to Metadata Handler on cache miss.
 	shared.NewCachingHandler(
```
Comment (Hexcles) [inline api/metadata_handler.go]: We shouldn't use CachingHandler any more, which adds yet another layer of memcache (keyed by URL) that's redundant now.

File: api/metadata_cache.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/web-platform-tests/wpt.fyi/shared"
+	"google.golang.org/appengine/memcache"
+)
+
+const metadataCacheKey = "WPT-METADATA"
+
+type webappMetadataFetcher struct {
+	ctx    context.Context
+	client *http.Client
+	log    shared.Logger
```
Comment (Hexcles) [inline api/metadata_cache.go]: Optional: we don't really need to have this field. `Fetch` can simply get a logger from `ctx` and pass it to `getMetadataFromMemcache`. This also works in tests: `TestContext` has a nil logger associated with it already.

File: api/query/metadata_cache.go [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+	"net/http"
+
+	"github.com/web-platform-tests/wpt.fyi/shared"
+)
+
+// MetadataMapCached is the local cache of Metadata in searchcache.
+var MetadataMapCached map[string][]byte = nil
+
+type searchcacheMetadataFetcher struct {
+	client *http.Client
```
Comment (Hexcles) [inline api/query/metadata_cache.go]: Constructing a `http.Client` is really cheap. I think we can construct a `&http.Client{Timeout: time.Second * 5}` directly in `Fetch` every time, which also has the extra benefit of having a single timeout definition.
